caravel/views.py

```diff
@@ -1843,25 +1846,41 @@ def csv(self, client_id):
     @expose("/fetch_datasource_metadata")
     @log_this
     def fetch_datasource_metadata(self):
-        # TODO: check permissions
-        # TODO: check if datasource exits
         session = db.session
         datasource_class = SourceRegistry.sources[request.args.get('datasource_type')]
         datasource = (
             session.query(datasource_class)
             .filter_by(id=request.args.get('datasource_id'))
             .first()
         )
-        # SUPPORT DRUID
-        # TODO: move this logic to the model (maybe)
-        grains_choices = [str(grain.name) for grain in datasource.database.grains()]
+        # Check if datasource exists
+        if not datasource:
+            return json_error_response(DATASOURCE_MISSING_ERR)
+        # Check permission for datasource
+        if not self.datasource_access(datasource):
+            return json_error_response(DATASOURCE_ACCESS_ERR)
+
+        time_columns = []
+        grains_choices = []
+        datasource_class_name = datasource_class.__name__
+        if datasource_class_name == 'SqlaTable':
+            time_columns = datasource.dttm_cols
+            grains = datasource.database.grains()
+            grains_choices = [grain.name for grain in grains]
+        elif datasource_class_name == 'DruidDatasource':
+            time_columns = ['all', '5 seconds', '30 seconds', '1 minute',
```

Comment:
do these time constants get used anywhere else? can we pull them to the top of the file as constants?